src/Computation/Context.cs

```diff
@@ -7,31 +8,77 @@ internal sealed class Context<TContextHandle> : IContext
     where TContextHandle : IContextHandle, new()
 {
     private readonly TContextHandle _contextHandle;
+    private readonly ISet<string> _names;
+    private readonly Lazy<Solver> _solver;
 
     public Context()
     {
         _contextHandle = new TContextHandle();
+        _names = new HashSet<string>();
+        _solver = new Lazy<Solver>(CreateSolver);
     }
 
     public void Dispose()
     {
         _contextHandle.Dispose();
     }
 
-    public Solver CreateSolver()
+    public void Assert(IEnumerable<BoolExpr> assertions)
     {
-        return _contextHandle.Context.MkSolver(_contextHandle.Context.MkTactic("smt"));
+        _solver.Value.Add(assertions);
+    }
+
+    public void Assert(string name, IEnumerable<BoolExpr> assertions)
+    {
+        if (_names.Contains(name))
+            return;
+
+        _names.Add(name);
+        Assert(assertions);
+    }
+
+    public Status Check(BoolExpr assertion)
+    {
+        return _solver.Value.Check(assertion);
+    }
+
+    public BitVecNum Evaluate(BitVecExpr variable)
+    {
+        return (BitVecNum) CreateModel(_solver.Value).Eval(variable, true);
+    }
+
+    public IEnumerable<KeyValuePair<FuncDecl, Expr>> Evaluate()
+    {
+        return CreateModel(_solver.Value).Consts;
     }
 
     public TSort CreateSort<TSort>(Func<Context, TSort> func)
         where TSort : Sort
     {
-        return func(_contextHandle.Context);
+        return Create(func);
     }
 
     public TExpr CreateExpr<TExpr>(Func<Context, TExpr> func)
         where TExpr : Expr
+    {
+        return Create(func);
+    }
+
+    private Solver CreateSolver()
+    {
+        return Create(c => c.MkSolver(c.MkTactic("smt")));
+    }
+
+    private TZ3Object Create<TZ3Object>(Func<Context, TZ3Object> func)
+        where TZ3Object : Z3Object
```

Comment:
I think we should drop this constraint and just call the generic type parameter `T` here as we don't use the constraint and the method is `private`.